servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/RoundRobinSelector.java

```diff
@@ -78,6 +104,92 @@ protected Single<C> selectConnection0(
 
     @Override
     public HostSelector<ResolvedAddress, C> rebuildWithHosts(@Nonnull List<? extends Host<ResolvedAddress, C>> hosts) {
-        return new RoundRobinSelector<>(index, hosts, getTargetResource(), failOpen);
+        return new RoundRobinSelector<>(index, hosts, getTargetResource(), failOpen, ignoreWeights);
+    }
+
+    private static Scheduler buildScheduler(AtomicInteger index, List<? extends Host<?, ?>> hosts) {
+        boolean allEqualWeights = true;
+        double maxWeight = 0;
+        for (Host<?, ?> host : hosts) {
+            double hostWeight = host.weight();
+            maxWeight = Math.max(maxWeight, hostWeight);
+            allEqualWeights = allEqualWeights && approxEqual(hosts.get(0).weight(), hostWeight);
+        }
+
+        if (allEqualWeights) {
+            return new ConstantScheduler(index, hosts.size());
+        } else {
+            double scaleFactor = MAX_WEIGHT / maxWeight;
+            int[] scaledWeights = new int[hosts.size()];
+
+            for (int i = 0; i < scaledWeights.length; i++) {
+                // Using ceil ensures a few things:
+                // - our max weighted element is picked on every round
+                // - hosts with weights near zero will never be truly zero
+                // - true zero weights will still be truly zero
+                scaledWeights[i] = Math.min(MAX_WEIGHT, (int) Math.ceil(hosts.get(i).weight() * scaleFactor));
+            }
+            return new StrideScheduler(index, scaledWeights);
+        }
+    }
+
+    private abstract static class Scheduler {
+        abstract int nextHost();
+    }
+
+    private static final class ConstantScheduler extends Scheduler {
+
+        private final AtomicInteger index;
+        private final int hostsSize;
+
+        ConstantScheduler(AtomicInteger index, int hostsSize) {
+            this.index = index;
+            this.hostsSize = hostsSize;
+        }
+
+        @Override
+        int nextHost() {
+            return (int) (Integer.toUnsignedLong(index.getAndIncrement()) % hostsSize);
```

Comment:
1. Consider keeping original `(index.getAndIncrement() & Integer.MAX_VALUE)` to avoid a cast from `long`.
2. Pre-existing, but if we are here wanna ask: when `index` overflows we loose the natural index increment. For example, if there are 13 hosts it will jump from idx=10 to idx=0 on overflow instead of going to inx=11. Should we worry about it or is it considered "ok" to interrupt a real round-robin once in a while?

Reply:
1. I like the `.toUnsignedLong` because it doubles the range an int can represent. The cast is clearly safe since `hostSize` is bounded to int sizes.
2. I'm not worried about a jump every ~4 billion requests.

Reply:
1. No worried about safety of the cast. If we see a trivial way to avoid unnecessary operation on the hot path, every small effort counts. I wouldn't say it's "premature optimization" bcz this is what we already had.
2. Ok.